### PR TITLE
H361: Direction-Magnitude decomposed WSS loss — WSS_z geometric attack

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Direction-Magnitude Decomposed WSS Loss** — decompose the WSS training gradient into (1) a cosine-similarity term that targets *direction* error and (2) a relative-norm term that targets *magnitude* error. The current per-channel MSE (even with `tau_z_loss_weight=1.67`) conflates these into a single scalar gradient per point, which underfits high-curvature vortex-shedding regions (wheel arches, door mirrors, A-pillar separation) where direction error and magnitude error have different difficulty/noise characteristics.

This is **structurally different** from the 3 closed scalar-reweight axes (H339 uniform, H341 wz-only, H346 per-vertex focal) — those tested scalar weights on existing MSE. H361 changes the loss **geometry** itself: cosine gradient pushes pred-direction → true-direction; relative-norm gradient pushes ‖pred‖ → ‖true‖. The two gradient fields are qualitatively different and can decompose a failure mode that a single scalar weight cannot.

**Mechanistic story**: WSS_z floor at 8.6122% has been falsified at every scalar reweight axis and every TTA axis. The remaining hypothesis is that the gradient signal itself is the wrong geometric object. Cosine + relative-magnitude is the natural decomposition for vector targets on a surface — used in vector field regression literature but never tried in this codebase.

**External evidence**: NVIDIA's GeoTransolver paper (arxiv:2512.20399) reports test_WSS=4.9% on DrivAerML — a 92bp delta over our 5.85% target. Their improvement comes from architectural changes (geometric cross-attention), which independently confirms WSS prediction is a representation-quality problem, not optimization. H361 is the loss-geometry version of the same hypothesis — fast to test, and serves as a clean discriminator: if H361 is null, the bottleneck is definitively architectural and we should escalate to GeoTransolver-style mods.

### Step 1 — implement loss function (train.py only)

Add after the existing `masked_mse` definition:

```python
def wss_direction_magnitude_loss(
    pred: torch.Tensor,    # [B, N, 3] — WSS channels τ_x, τ_y, τ_z only
    target: torch.Tensor,  # [B, N, 3]
    mask: torch.Tensor,    # [B, N]
    alpha: float = 1.0,
    beta: float = 1.0,
    eps: float = 1e-6,
) -> torch.Tensor:
    """Decomposed WSS loss: alpha · (1 - cos_sim) + beta · relative_norm_mse."""
    dot = (pred * target).sum(-1, keepdim=True)          # [B, N, 1]
    pred_norm = pred.norm(dim=-1, keepdim=True).clamp(min=eps)
    tgt_norm  = target.norm(dim=-1, keepdim=True).clamp(min=eps)
    cos_sim = dot / (pred_norm * tgt_norm)               # [B, N, 1]
    dir_loss = masked_mean((1.0 - cos_sim).squeeze(-1), mask)
    mag_sq_err = (pred_norm.squeeze(-1) - tgt_norm.squeeze(-1)).square()
    rel_mag_loss = masked_mean(mag_sq_err / (tgt_norm.squeeze(-1).square() + eps), mask)
    return alpha * dir_loss + beta * rel_mag_loss
```

### Step 2 — Config flags (train.py Config dataclass)

```python
wss_dir_loss_alpha: float = 0.0   # 0.0 = disabled (control)
wss_dir_loss_beta: float = 0.0    # 0.0 = disabled (control)
```

### Step 3 — Wire into train_loss

After `surface_loss = masked_mse(...)` (keep this intact, including `tau_z_loss_weight=1.67`):

```python
if config.wss_dir_loss_alpha > 0.0 or config.wss_dir_loss_beta > 0.0:
    wss_pred = out["surface_preds"][..., 1:4]   # τ_x, τ_y, τ_z (verify channel layout in model.py before running)
    wss_true = surface_target[..., 1:4]
    wss_extra = wss_direction_magnitude_loss(
        wss_pred, wss_true, batch.surface_mask,
        alpha=config.wss_dir_loss_alpha,
        beta=config.wss_dir_loss_beta,
    )
    surface_loss = surface_loss + wss_extra
```

**Verify channel layout first.** In `model.py` confirm `surface_preds[..., 0]` = cp and `surface_preds[..., 1:4]` = τ_x, τ_y, τ_z. If different, slice accordingly. **Do NOT modify model.py** — read-only sanity check only.

### Step 4 — Smoke (5 min, sanity wiring)

```bash
torchrun --nproc_per_node=8 train.py \
  --epochs 1 --agent edward \
  --wandb_name "edward/h361-smoke" \
  --wandb_group h361-wss-dirloss \
  --wss_dir_loss_alpha 0.5 --wss_dir_loss_beta 0.5 \
  --tau_z_loss_weight 1.67 \
  --train_surface_points 8000 --eval_surface_points 8000 \
  --train_volume_points 8000 --eval_volume_points 8000 \
  --kill-thresholds "20:train/loss<1000"
```

Expect: smoke completes ~5min, `train/loss` finite and decreasing.

### Step 5 — Phase 1 Arm A (3-epoch fine-tune from H336 EP13, ~6.5h)

Find the H336 EP13 run ID (you used this in H312 and H338 — same checkpoint resume). Then:

```bash
torchrun --nproc_per_node=8 train.py \
  --epochs 16 --epochs_already_done 13 \
  --agent edward \
  --wandb_name "edward/h361-armA-a0.5b0.5" \
  --wandb_group h361-wss-dirloss \
  --resume_from_wandb <H336_EP13_RUN_ID> --resume_alias epoch-13 \
  --lr 1e-4 --lr_cosine_t_max 3 --lr_min 1e-6 \
  --tau_z_loss_weight 1.67 \
  --wss_dir_loss_alpha 0.5 --wss_dir_loss_beta 0.5 \
  --kill-thresholds "500:train/loss<5,2000:val_primary/abupt_axis_mean_rel_l2_pct<25"
```

### Step 6 — TTA + cal eval (identical to H342)

K=5, Student-t ν=4, σ=5e-4, mirror, 8 resolutions 32k→131k, per-channel OLS cal. Use the same eval script you used for H338 / H312. Report val_abupt_calibrated, test_abupt_calibrated, full per-axis breakdown especially test_WSS_z.

### Step 7 — Phase 1 Arm B (conditional, only if Arm A val_raw < 5.93)

Direction-dominant: `--wss_dir_loss_alpha 1.0 --wss_dir_loss_beta 0.25`. Same Phase 1 + TTA cascade. ~6.5h.

## Pre-committed close rules

Close immediately if ANY:
- Smoke: `train/loss` NaN or >10 after 20 steps
- Phase 1 EP14: `val_primary/abupt_axis_mean_rel_l2_pct > 6.05` (>10bp above SOTA raw — cal cannot rescue per `cal-cannot-rescue-train-raw-regression`)
- Phase 1 final EP16: `val_raw > 5.98` (no improvement over H342 baseline raw)
- Phase 1 final EP16: `val_raw 5.93-5.98` → run TTA on val only; if val_cal misses gate by >5bp, close (don't burn test eval)

**Proceed to full TTA+test eval only if Phase 1 val_raw < 5.93.**

## Baseline

Current SOTA: **H342 3-cp output-avg ep13+ep14+ep15 × K=4 TTA (PR #1526 MERGED)**

| Metric | H342 SOTA | Merge gate |
|---|---:|---:|
| **val_abupt_calibrated** | **5.8962%** | < **5.8962%** |
| **test_abupt_calibrated** | **5.7357%** | < **5.7357%** |
| test_WSS_z_cal | 8.6122% | (target obstacle, Transolver-3 floor 5.85%) |
| test_VP_cal | 3.3751% | (safely under 3.421 floor) |
| test_SP_cal | 3.6124% | (over 3.577 floor by +3.5bp — secondary) |

**Merge gate**: val_cal < **5.8962%** AND test_cal < **5.7357%** (AND-logic)

**W&B run IDs (reference)**:
- H342 SOTA evals: 3icmxaqe (ep15), qgw0ix77 (ep14), ijadzof0 (ep13)
- H336 EP13 source checkpoint: look up `nezuko/h336-*` runs in W&B (you've used these as source in prior experiments)

## Constraints (REINFORCED)

- **z-axis τ_z LOCKED at trained value 1.67** — do NOT modify `tau_z_loss_weight`
- **Read-only files**: `data/loader.py`, `data/preload.py`, `data/split_manifest.json`
- **DDP 8-GPU mandatory** — every training run uses `torchrun --nproc_per_node=8`
- **SENPAI_TIMEOUT_MINUTES=540** hard cap
- **No param overhead** — H361 is loss-only, zero new params, must remain so
- **No ensembles** (lazy route, Morgan)
- One terminal `SENPAI-RESULT` marker required for merge consideration

## Reproduce command (for advisor verification)

```bash
cd target && torchrun --nproc_per_node=8 train.py \
  --epochs 16 --epochs_already_done 13 --agent edward \
  --wandb_name "edward/h361-armA-a0.5b0.5" \
  --wandb_group h361-wss-dirloss \
  --resume_from_wandb <H336_EP13_RUN_ID> --resume_alias epoch-13 \
  --lr 1e-4 --lr_cosine_t_max 3 --lr_min 1e-6 \
  --tau_z_loss_weight 1.67 \
  --wss_dir_loss_alpha 0.5 --wss_dir_loss_beta 0.5
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR heartbeat 2026-06-02 01:35Z: H361 Phase 1 Arm A (α=0.5 β=0.5) confirmed RUNNING healthy via W&B (`zjl03wit` etc.). All 8 ranks at step 31, rt ~135s, train/loss 0.013-0.018 — well-behaved. Smoke crashes resolved. Phase 1 ETA ~04:30Z (3h training). Continue training; post SENPAI-RESULT when terminal.

**Comment by morganmcg1** [comment]
ADVISOR heartbeat 2026-06-02 03:45Z: clearing harness stale_wip flag.

**PR #1553 H361 Arm A α=0.5/β=0.5 status:**
- Phase 1 training: rank-0 `zjl03wit` at **step 8163 rt 119min, train/loss 0.012**
- All 8 ranks DDP healthy (step 8580+, train_loss 0.003-0.017 across ranks)
- Phase 1 should land terminal at EP16 within next ~5-10 min

Standing by for SENPAI-RESULT after Phase 1 eval. Per the original PR plan, you'll then make the Arm B/C / TTA-cascade decision based on Phase 1 raw close-rule check.

No action needed right now.

**Comment by morganmcg1** [comment]
STUDENT edward:
SENPAI-RESULT: {"terminal":true,"status":"close","pending_arms":false,"wandb_run_ids":["zjl03wit"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":6.2809}}

## Results — H361 Arm A α=0.5 β=0.5 — CLOSED (null)

**Decision**: Close. Phase 1 raw val_primary tripped BOTH pre-committed close rules; H342 cal SOTA gate cannot be reached from a +30bp raw regression. Arm B is gated on `Arm A val_raw < 5.93%` → **not met**, so Arm B will NOT run.

### Per-epoch val_primary trajectory (raw, no TTA)

| Epoch | abupt% | WSS_x | WSS_y | WSS_z | WSS_vec | SP | VP | train_loss |
|------:|------:|------:|------:|------:|------:|------:|------:|------:|
| EP13 (source H336, yw2a5dyl) | **6.017** | 5.942 | 7.442 | 9.194 | — | 3.967 | 3.541 | — |
| EP14 | 6.703 | 6.601 | 8.178 | 9.848 | 7.475 | 4.442 | 4.447 | 0.0328 |
| EP15 | 6.344 | 6.214 | 7.809 | 9.457 | 7.097 | 4.232 | 4.008 | 0.0191 |
| **EP16 (best)** | **6.281** | 6.130 | 7.731 | **9.397** | 7.020 | 4.200 | 3.947 | **0.0153** |

### Close-rule trigger

| Rule | Threshold | Actual | Verdict |
|---|---:|---:|---|
| EP14 close rule | val_raw > 6.05% | 6.703% | **TRIPPED (+65bp)** |
| EP16 close rule | val_raw > 5.98% | 6.281% | **TRIPPED (+30bp)** |
| Arm B gate | val_raw < 5.93% | 6.281% | **NOT MET** |

### What happened

The decomposed cosine+rel-magnitude WSS term degraded the model on every channel from the EP13 starting point — not just WSS_z which it was designed to attack. After 3 epochs of fine-tune with the H342 LR schedule:

- **WSS_z**: 9.194 → 9.397 (**+20bp worse**) — the targeted axis got worse, not better
- **WSS_x**: 5.942 → 6.130 (+19bp)
- **WSS_y**: 7.442 → 7.731 (+29bp)
- **SP**: 3.967 → 4.200 (+23bp)
- **VP**: 3.541 → 3.947 (+41bp)
- **abupt (primary)**: 6.017 → 6.281 (+26bp)

train_loss still trended down (0.033 → 0.019 → 0.015), but it is a different loss object (added cosine + relative-magnitude terms), so train_loss is no longer comparable across H336 vs H361. The base surface MSE was untouched — adding the cosine/relmag gradient pulled the optimizer toward solutions that minimize the new combined loss at the cost of every per-channel rel_l2 the benchmark cares about.

**Mechanistic read**: Adding `(1 - cos_sim) + relative-norm-mse` to a model already at convergence on per-channel MSE biases gradients away from the rel_l2 minimum. The cosine term rewards direction-matching at small-norm points (which are noise on the surface) just as strongly as at flow-separated high-norm points. The relative-norm term over-weights small-magnitude regions (denominator `||target||²` is tiny). Net effect: the loss landscape near the H336 optimum tilts toward direction/magnitude on irrelevant low-WSS surface, and the optimizer escapes the H336 minimum into a worse rel_l2 basin.

**Discriminator outcome**: PR called this out — "if H361 is null, the bottleneck is definitively architectural and we should escalate to GeoTransolver-style mods." H361 is null. **WSS_z floor is not a loss-geometry problem; the per-channel MSE gradient is already the right object — the bottleneck is representation.**

### Exact command used

```bash
cd target && torchrun --nproc_per_node=8 train.py \
  --epochs 16 --epochs_already_done 13 --agent edward \
  --wandb_name "edward/h361-armA-a0.5b0.5" --wandb_group h361-wss-dirloss \
  --resume_from_wandb yw2a5dyl --resume_alias epoch-13 \
  --lr 1e-4 --lr_cosine_t_max 3 --lr_min 1e-6 \
  --tau_z_loss_weight 1.67 \
  --wss_dir_loss_alpha 0.5 --wss_dir_loss_beta 0.5 \
  --kill-thresholds "500:train/loss<5,2000:val_primary/abupt_axis_mean_rel_l2_pct<25"
```

### Run metadata

- W&B run ID (rank-0): `zjl03wit` ("edward/h361-armA-a0.5b0.5-rank0")
- DDP 8-GPU, peak val/pre_clear_reserved ≈ 67.9 GB per GPU
- Wall time: ~131 min (8163 steps × ~135s/step DDP across 8 ranks)
- Source checkpoint: `yw2a5dyl` (H336 EP13, ema)

### Decision on Arm B (α=1.0 β=0.25)

**Skip.** Arm B was explicitly conditional on Arm A val_raw < 5.93%. Arm A is +35bp away (6.281%). Direction-dominant weighting won't fix the underlying mechanism — the cosine gradient is the thing biasing the optimizer away from the rel_l2 minimum, and Arm B *increases* the cosine weight. Spending another 6.5h on it is negative expected value.

### Suggested follow-ups

1. **Escalate to architectural attacks on WSS_z** (the discriminator outcome of H361). GeoTransolver-style geometric cross-attention is the obvious next step — its 4.9% WSS on DrivAerML is +92bp over our target. Lower-cost intermediate: try a WSS-specific decoder head (separate MLP for WSS τ_x/τ_y/τ_z conditioned on local surface curvature features) without changing the backbone — single-PR experiment.
2. **Per-point-mass scalar reweight** (closes the per-channel-MSE family before architectural moves). PR family H339/H341/H346 explored point-mass weights on existing channel grids; the unexplored axis is `||target||`-conditioned per-point weight (up-weight only when local WSS magnitude is in the top decile — vortex-shedding hotspots). Cheap, loss-only, structurally different from cosine.
3. **Loss with target-normalized rel_l2 directly** instead of decomposing. The benchmark IS rel_l2 — train with per-channel rel_l2 (sum of squared errors / sum of squared targets per case, masked) as the loss. This is the *aligned* training objective, not a decomposition of it. Could be one PR.

The H361 close result is mechanistically informative: it suggests the right next move is architectural (point 1) or aligned-objective training (point 3), not further reweighting of the existing MSE.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1553 because ADVISOR 2026-06-02 04:05Z: closing PR #1553 H361 as null.

**Accepted close — clean discriminator outcome.**

Your pre-committed close-rules tripped emphatically (EP14 +65bp, EP16 +30bp). val_raw 6.281 is +35bp from the Arm B gate; running Arm B is negative expected value as you correctly identified. WSS_z went 9.194 → 9.397 — the targeted axis got worse, not better.

**Finding L (12th decoder/input/loss-geometry axis closed)**: `wss-direction-magnitude-decomposed-loss-null`. The mechanism you identified is exactly right — cosine `(1−cos_sim)` rewards direction-matching at all magnitudes including noise; relative-norm `MSE/||target||²` over-weights small-magnitude regions. Both gradient fields pull the optimizer AWAY from rel_l2 minimum at the H336 convergence point. Per-channel MSE *is* the right gradient object for this benchmark.

**Discriminator signal**: This was designed to discriminate "WSS_z floor is loss-geometry" vs "representation problem". You confirmed representation. **The H361 close + H357 GeoTransolver-content-embedding close (this loop) together rule out content-injection and loss-decomposition as easy WSS_z attacks.** Architectural escalation is now the explicit next move.

**Bank** of your suggested follow-ups for the next research planning round:
1. ||target||-magnitude-conditioned per-point weight (vortex-shedding hotspots, top-decile WSS reweighting) — fresh, cheap, single-PR
2. Direct rel_l2 loss training (aligned objective, sum_sq_err / sum_sq_target per case)
3. GeoTransolver-style geometric cross-attention (longer-form architectural)

These get queued for the next round of hypothesis generation.

Nice rigorous discriminator design. Closing..
